dbx/sync/event_handler.py

```diff
@@ -0,0 +1,90 @@
+import threading
+from contextlib import contextmanager
+from typing import List
+
+from watchdog.events import FileSystemEvent, FileSystemEventHandler
+from watchdog.observers import Observer
+
+from .path_matcher import PathMatcher
+
+
+class CollectingEventHandler(FileSystemEventHandler):
+    """A watchdog event handler that collects all the events captured so they can be retrieved in batches."""
+
+    def __init__(self, *, matcher: PathMatcher = None):
+        super().__init__()
+
+        self.lock = threading.RLock()
+        self.events = []
+        self.matcher = matcher
+
+    def _should_ignore(self, event: FileSystemEvent) -> bool:
+        return not self.matcher.match(event.src_path, is_directory=event.is_directory) if self.matcher else False
+
+    def get_events(self) -> List[FileSystemEvent]:
+        """Gets the events collected since the last time this method was called.
+
+        Returns:
+            List[FileSystemEvent]: events collected since the last call
+        """
+        with self.lock:
```

Comment:
it's an interesting combination of using locks and asyncio ;) i'd expect a channel here or `asyncio.Queue` and confinement of event handling to a single loop to avoid locks.

Reply:
The event handler is used alongside the async calls but not within them.  If you look in `main_loop`, it uses `event_handler.get_events` to check for any events, then when there are some it uses `asyncio.run`.